pennylane/devices/default_tensor.py

```diff
@@ -889,3 +911,113 @@ def execute_and_compute_vjp(
+    pw = next(
+        iter(
+            qml.pauli.string_to_pauli_word(
+                ops._hyperparameters["pauli_word"], wire_map
+            ).pauli_rep.keys()
+        )
+    )
+    arrays = []
+    sites = list(pw.keys())
+    for i, P in enumerate(pw.values()):
+        if i == 0:
+            arr = qml.math.zeros((1, 2, 2, 2), dtype=complex)
+            arr[0, 0] = _PAULI_MATRICES[P]
+            arr[0, 1] = qml.math.eye(2, dtype=complex)
+
+        elif i == len(wire_map) - 1:
+            arr = qml.math.zeros((2, 1, 2, 2), dtype=complex)
+            arr[0, 0] = _PAULI_MATRICES[P] * (-1j) * qml.math.sin(theta / 2)
+            arr[1, 0] = qml.math.eye(2, dtype=complex) * qml.math.cos(theta / 2)
+
+        else:
+            arr = qml.math.zeros((2, 2, 2, 2), dtype=complex)
+            arr[0, 0] = _PAULI_MATRICES[P]
+            arr[1, 1] = qml.math.eye(2, dtype=complex)
+
+        arrays.append(arr)
+
+    mpo = qtn.MatrixProductOperator(arrays=arrays, sites=sites)
+    device._quimb_circuit._psi = mpo.apply(
+        device._quimb_circuit.psi, max_bond=device._max_bond_dim, cutoff=device._cutoff
+    )
+
+
+@apply_operation_core.register
+def apply_operation_core_trotter_product(ops: qml.TrotterProduct, device):
+    """Dispatcher for _apply_operation."""
+    time = ops.data[-1]
+    n = ops._hyperparameters["n"]
+    order = ops._hyperparameters["order"]
+    ops = ops._hyperparameters["base"].operands
+    decomp = _recursive_expression(time / n, order, ops)[::-1] * n
+    for o in decomp:
+        device._quimb_circuit.apply_gate(
+            qml.matrix(o).astype(device._c_dtype), *o.wires, parametrize=None
+        )
+
+
+@singledispatch
+def expval_core(obs: Observable, device) -> float:
+    """Dispatcher for expval."""
+    return device._local_expectation(qml.matrix(obs), tuple(obs.wires))
+
+
+@expval_core.register
+def expval_core_tensor(obs: Tensor, device) -> float:
+    """Computes the expval of a Tensor."""
+    return expval_core(Prod(*obs._args), device)
+
+
+@expval_core.register
+def expval_core_prod(obs: Prod, device) -> float:
+    """Computes the expval of a Prod."""
+    ket = copy.deepcopy(device._quimb_circuit)
+    for op in obs:
+        ket.apply_gate(qml.matrix(op).astype(device._c_dtype), *op.wires, parametrize=None)
+    return np.real((device._quimb_circuit.psi.H & ket.psi).contract(all, output_inds=()))
+
+
+@expval_core.register
+def expval_core_sprod(obs: SProd, device) -> float:
+    """Computes the expval of a SProd."""
+    return obs.scalar * expval_core(obs.base, device)
+
+
+@expval_core.register
+def expval_core_sum(obs: Sum, device) -> float:
+    """Computes the expval of a Sum."""
+    return sum(expval_core(m, device) for m in obs)
+
+
+@expval_core.register
+def expval_core_linear_combination(obs: LinearCombination, device) -> float:
+    """Computes the expval of a LinearCombination."""
+    return sum(expval_core(m, device) for m in obs)
```

Comment:
I think it boils down to the same thing no (each term will be split in `expval_core_sprod`)?

Reply:
Yes but I think its faster to compute the matrix of just the op and then multiply rather than directly the matrix of the sprod:

```python
sprod = qml.s_prod(0.5, qml.X(0))
op = qml.X(0)

%timeit qml.matrix(sprod)
%timeit 0.5 * qml.matrix(op)
```

```pycon
8.12 µs ± 28.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
2.3 µs ± 20 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Probably in the bigger picture of things no big difference but for LinearCombination that is kind of the point to be able to directly access its `terms`

Reply:
This is what `expval_core_sprod` does, so we're not computing the matrix of the `sprod`.

Reply:
ah you're right, thanks! marking as resolved :)